### PR TITLE
Hive: Implement Deserializer for Hive writes

### PR DESCRIPTION
**Pull request by pvary**
Implements the Deserializer for Hive writes, and creates the corresponding test suite

### CODE REVIEW

**Comment by pvary** [comment]
@marton-bod, @lcspinter: Could you please review?
Thanks,
Peter

File: mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspector.java [hunk 1]
```diff
@@ -22,26 +22,38 @@
...
+    implements TimestampObjectInspector, IcebergWriteObjectInspector {
 
   private static final IcebergTimestampObjectInspector INSTANCE_WITH_ZONE = new IcebergTimestampObjectInspector() {
     @Override
     LocalDateTime toLocalDateTime(Object o) {
       return ((OffsetDateTime) o).toLocalDateTime();
     }
+
+    @Override
+    public Object getIcebergObject(Object o) {
```
Comment (marton-bod) [inline mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspector.java:41]: Can't we return an `OffsetDateTime` object?

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.StringObjectInspector;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.mr.hive.serde.objectinspector.IcebergWriteObjectInspector;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+class Deserializer {
+  private FiledDeserializer mainDeserializer;
```
Comment (marton-bod) [inline mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:42]: nit: unless we have other deserializers, I would just name this field `deserializer` or `fieldDeserializer`

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      case LONG:
+        return o -> ((LongObjectInspector) fieldInspector).get(o);
+      case FLOAT:
+        return o -> ((FloatObjectInspector) fieldInspector).get(o);
+      case DOUBLE:
+        return o -> ((DoubleObjectInspector) fieldInspector).get(o);
+      case STRING:
+        return o -> ((StringObjectInspector) fieldInspector).getPrimitiveJavaObject(o);
+      case UUID:
+        // TODO: This will not work with Parquet. Parquet UUID expect byte[], others are expecting UUID
```
Comment (marton-bod) [inline mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:71]: Is there a way to handle the Parquet case? Just wondering what happens if someone tries to write UUID data with parquet

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+      if (object == null) {
+        return null;
+      }
+
+      List<Object> data = fieldInspector.getStructFieldsDataAsList(object);
+      Record result = GenericRecord.create(type);
+
+      for (int i = 0; i < filedDeserializers.length; i++) {
+        Object fieldValue = data.get(i);
+        if (fieldValue != null) {
```
Comment (marton-bod) [inline mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:119]: if the fieldValue is null, shouldn't we still set the null in the result record?

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  private static final StandardStructObjectInspector CUSTOMER_OBJECT_INSPECTOR =
+      ObjectInspectorFactory.getStandardStructObjectInspector(
+          Arrays.asList("customer_id", "first_name"),
+          Arrays.asList(
+              PrimitiveObjectInspectorFactory.writableLongObjectInspector,
+              PrimitiveObjectInspectorFactory.writableStringObjectInspector
+          ));
+
+
```
Comment (marton-bod) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java:54]: Can we add a test case where a `SerdeException` should be thrown?

**Review by marton-bod — COMMENTED** [review]
generally looks good, just a few minor questions

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+  Deserializer(Schema schema, ObjectInspector fieldInspector) throws SerDeException {
+    this.mainDeserializer = deserializer(schema.asStruct(), fieldInspector);
+  }
+
+  Record deserialize(Object data) {
+    return (Record) mainDeserializer.value(data);
+  }
+
+  private interface FiledDeserializer {
```
[the remark this answers is not in the document]
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java]: Thanks for catching!
Fixed

File: mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspector.java (hunk 1 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspector.java]: Fixed

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java (hunk 2 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java]: Renamed to fieldDeserializer

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java (hunk 4 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:119]: We created a new Record here, so it should be null by default.
That said it should not hurt too much on performance, and easier to read when this is explicit, so added the extra set

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java (hunk 5 above)
Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java:56]: Added the extra case

**Comment by pvary** [comment]
@rdblue, @shardulm94: Could you please review if you have some time? This is a part of #1407 (Hive: HiveIcebergOutputFormat first implementation for handling Hive inserts into unpartitioned Iceberg tables)

Thanks!
Peter

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java (hunk 3 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:71]: I think we should find a general way for handling UUID's across different file formats. This is a broader question so I would leave this as it is for this PR, and fix in some upcoming one

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+      case DECIMAL:
+        // Iceberg specific conversions
+        return o -> ((IcebergWriteObjectInspector) fieldInspector).getIcebergObject(o);
+      case STRUCT:
+        return new StructDeserializer((Types.StructType) type, (StructObjectInspector) fieldInspector);
+      case LIST:
+      case MAP:
+      case TIME:
+      default:
+        throw new SerDeException("Unsupported column type: " + type);
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:86]: List and Map aren't supported?

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+  Record deserialize(Object data) {
+    return (Record) fieldDeserializer.value(data);
+  }
+
+  private interface FieldDeserializer {
+    Object value(Object object);
+  }
+
+  private static FieldDeserializer deserializer(Type type, ObjectInspector fieldInspector) throws SerDeException {
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:56]: Most of the other modules use the visitor pattern to traverse a type. That keeps the logic for traversing a schema in just one place so you don't need to mix it in with your domain-specific code.

It looks like this method is called from with the `StructDeserializer` constructor, so there is a recursive traversal of the schema that goes back and forth between object constructors and this method. That's a bit hard to follow, so I think it would be simpler if you took the visitor approach.

A `GenericAvroReader` is similar to what you're building here, so take a look at that as an example: https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/avro/GenericAvroReader.java

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    private StructDeserializer(Types.StructType type, StructObjectInspector fieldInspector) throws SerDeException {
+      List<? extends StructField> structFields = fieldInspector.getAllStructFieldRefs();
+      List<Types.NestedField> nestedFields = type.fields();
+      this.fieldDeserializers = new FieldDeserializer[structFields.size()];
+      this.fieldInspector = fieldInspector;
+      this.type = type;
+
+      for (int i = 0; i < fieldDeserializers.length; i++) {
+        fieldDeserializers[i] =
+            deserializer(nestedFields.get(i).type(), structFields.get(i).getFieldObjectInspector());
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:104]: As I noted above, I think this would be cleaner if you used a visitor and passed the field deserializers into this constructor instead of calling deserializer here.

File: mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergWriteObjectInspector.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.mr.hive.serde.objectinspector;
+
+public interface IcebergWriteObjectInspector {
+  Object getIcebergObject(Object value);
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergWriteObjectInspector.java:23]: We avoid `get` in APIs because it doesn't add much value and could be misleading in some cases. I think it is misleading here because this actually converts the object to the value to the Iceberg generic representation rather than just returning it. We also shouldn't need "Iceberg" in the name because it's in `IcebergWriteObjectInspector`.

How about naming this `convert`?

Also, where we have similar interfaces elsewhere, we tend to like to use parameterized types. [`ValueWriters`](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/avro/ValueWriters.java#L190-L200) is a good example.

File: mr/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+    record.set(4, 4.2d);
+    record.set(5, LocalDate.of(2020, 1, 21));
+    // TimeType is not supported
+    // record.set(6, LocalTime.of(11, 33));
+    // Nano is not supported ?
+    record.set(6, OffsetDateTime.of(2017, 11, 22, 11, 30, 7, 0, ZoneOffset.ofHours(2)));
+    record.set(7, LocalDateTime.of(2019, 2, 22, 9, 44, 54));
+    record.set(8, "kilenc");
+    if (uuidAsByte) {
+      // TODO: Parquet UUID expect byte[], others are expecting UUID
```
Comment (rdblue) [inline mr/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java:117]: Can you file a bug for this?

File: mr/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+    for (int i = 0; i < HiveIcebergTestUtils.FULL_SCHEMA.columns().size(); i++) {
+      record.set(i, null);
+    }
+
+    return record;
+  }
+
+  public static List<Object> valuesForTestRecord(Record record) {
+    ByteBuffer byteBuffer = record.get(11, ByteBuffer.class);
+    byte[] bytes = new byte[byteBuffer.remaining()];
```
Comment (rdblue) [inline mr/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java:141]: You should be able to use utils in `ByteBuffers` for this.

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+  );
+
+  private static final StandardStructObjectInspector CUSTOMER_OBJECT_INSPECTOR =
+      ObjectInspectorFactory.getStandardStructObjectInspector(
+          Arrays.asList("customer_id", "first_name"),
+          Arrays.asList(
+              PrimitiveObjectInspectorFactory.writableLongObjectInspector,
+              PrimitiveObjectInspectorFactory.writableStringObjectInspector
+          ));
+
```
Comment (rdblue) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java:53]: Nit: double whitespace.

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    expected.set(1, "Bob");
+
+    Record actual = deserializer.deserialize(new Object[] { new LongWritable(1L), new Text("Bob") });
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void testDeserializeEverySupportedType() throws SerDeException {
+    // No test yet for Hive3 (Date/Timestamp creation)
```
Comment (rdblue) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java:70]: I think that Assume can be passed string context. This would be a good thing to pass in.

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+    Record actual = deserializer.deserialize(nulls);
+
+    Assert.assertEquals(expected, actual);
+
+    // Check null record as well
+    Assert.assertNull(deserializer.deserialize(null));
+  }
+
+  @Test(expected = SerDeException.class)
```
Comment (rdblue) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java:100]: We typically prefer to use `AssertHelpers.assertThrows` rather than `expected`. When we're testing other places, we can continue making assertions after the exception is thrown that way. Here it probably doesn't matter much, but it is a good habit to be in.

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+  public void testSerDeException() throws SerDeException {
```
Comment (rdblue) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java:101]: Why did you choose to use `SerDeException`? Is that expected by Hive?

**Comment by rdblue** [comment]
Thanks, @pvary! This looks great so far. I had a few comments, but it looks like a good improvement!

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java (hunk 7 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java]: Haven't planned for that but it was quite easy with the Visitor, so I have done this.

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java (hunk 8 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java]: Ok. I moved to the Visitor pattern. The the code is quite nice, and compact 😄 

Hive struct names are not matched to the Iceberg struct name fields in case of the main struct which contains the result columns of a query. In this case we have the following col names: `0:col1`, `1:col2` and they should be matched by the positions of the fields and not by the names.

Handling this will add more complexity. I propose to handle this in the next PR. What do you think?

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java (hunk 9 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java]: Done

File: mr/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java (hunk 12 above)
Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java]: Done.
Thanks!

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java (hunk 13 above)
Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java:56]: Shall we use 2 spaces for `Continuation indent` as well?

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java (hunk 14 above)
Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java]: Thanks for noticing!
Done

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java (hunk 15 above)
Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java]: SerDeException was not a runtime exception and would have to do some extra magic to use `AssertHelpers.assertThrows`.

On the other hand I think we can convert any exception in the SerDe to SerDeException, we can throw `IllegalArgumentException`s instead and handle those outside.

Fixed

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java (hunk 16 above)
Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java]: Yeah. The SerDe should only throw SerDeException.
But we can throw whatever we want here, and wrap it later to SerDeException.

Fixed

File: mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergWriteObjectInspector.java (hunk 10 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergWriteObjectInspector.java]: My reasoning was:
- This is an `ObjectInspector` only change and adds only a single new method to the Iceberg specific` ObjectInspector`s
- All of the `ObjectInspector` implementations are Overriding `get*` methods as inherited from Hive
- It might be reasonable to stick to one naming methods on Object level

That said, since I am not overly attached to the original naming schema, changed it to the suggested one.

Thanks,
Peter

Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergWriteObjectInspector.java]: Oh.. and I plan to collapse the `DeserializerVisitor.primitive` to the following when we have implemented the handling of every Iceberg primitive type:
```
    @Override
    public FieldDeserializer primitive(PrimitiveType type, ObjectInspector inspector) {
      if (inspector instanceof WriteObjectInspector) {
        return o -> ((WriteObjectInspector) inspector).convert(o);
      } else {
        return o -> ((PrimitiveObjectInspector) inspector).getPrimitiveJavaObject(o);
      }
    }
```
Having differently named method for every type would make this impossible.

File: mr/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java (hunk 11 above)
Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java:118]: Filed an issue for this: #1881

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java (hunk 13 above)
Comment (rdblue) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java:56]: Continuation indents are 2 indents, which are 4 spaces. What you have here looks correct to me.

Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java:56]: Then I am not sure what was the comment about. 😢

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+    }
+
+    @Override
+    public FieldDeserializer list(ListType listTypeInfo, ObjectInspector inspector, FieldDeserializer deserializer) {
+      return o -> {
+        if (o == null) {
+          return null;
+        }
+
+        List<Object> result = new ArrayList<>();
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:159]: Nit: prefer `Lists.newArrayList()`.

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+        for (Map.Entry<?, ?> entry : mapObjectInspector.getMap(o).entrySet()) {
+          result.put(keyDeserializer.value(entry.getKey()), valueDeserializer.value(entry.getValue()));
+        }
+        return result;
+      };
+    }
+  }
+
+  private static class PartnerObjectInspectorByNameAccessors
+      implements SchemaWithPartnerVisitor.PartnerAccessors<ObjectInspector> {
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:190]: Minor: This could probably be a singleton.

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+    Object value(Object object);
+  }
+
+  /**
+   * Hive query results schema column names do not match the target Iceberg column names.
+   * Instead we have to rely on the column order. To keep the other parts of the code generic we fix this with a
+   * wrapper around the ObjectInspector. This wrapper uses the Iceberg schema column names instead of the Hive column
+   * names for {@link #getStructFieldRef(String) getStructFieldRef}
+   */
+  private static class FixNameMappingObjectInspector extends StructObjectInspector {
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:227]: Don't all of the structs need to be wrapped by this?

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
I don't have any major concerns about this, so I'm going to merge it to unblock the other work. Thanks @pvary!

It would be nice to figure out a better way to traverse the object inspector tree that doesn't require the `FixNameMappingObjectInspector`. I'm not quite sure what that is, but I think it might be a matter of converting the name into a position and accessing the object inspectors by position. That can be done later, though.

**Comment by pvary** [comment]
Thanks @rdblue for the review and the merge!

In Hive you always need to provide the name-value pair for structs. AFAIK the only exception is the Schema level struct.
I have been thinking about 2 other solutions:
- Move field matching to be based on the ordering of the fields. We can do it by adding a new input to parameter to `SchemaWithPartnerVisitor<P, R>.PartnerAccessors<P>.fieldPartner(P partnerStruct, int fieldId, String name, int fieldOrderId)`. I do not like this because I do not feel that Struct fields should have specific ordering. It could work, because we are the ones who generate the ObjectInspectors, but still...
- Add the field name to the ObjectInspectors themselves when generating them for the Schema. I do not like this because then we will have specific ObjectInspector instances we have to create and drop.

So this is why I have kept the current one which has its own drawbacks, so I am happy to go for a more general solution if we find one.

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java (hunk 17 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:159]: I remember on some other PR got a review from someone else to avoid unnecessary guava uses.
I am find with both, but I think we should stick to one or the other. Shall it be Lists.newArrayList then?

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java (hunk 13 above)
Comment (rdblue) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java:56]: I think it was an extra newline.

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java (hunk 17 above)
Comment (rdblue) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:159]: We don't want to bring in any additional Guava classes if possible, but using the ones that are already there is a good thing. For this case, we can easily replace map class implementations with an import change later.
